[WEBSITE-666] - Hacktoberfest: Update the /participate/code page (#2538)
jenkins-infra/jenkins.io: content/participate/code.adoc

---
layout: project
title: Contribute code
section: participate
tags:
  - outreach-programs
  - developer
---

Jenkins project includes a lot of code, and we invite everyone to contribute to the project.
There is a diverse set of programming languages used in Jenkins,
including but not limited to: Java, JavaScript, Groovy, Golang, Ruby, Shell scripts.
And, since Jenkins is an automation server with hundreds of plugins, there is a huge number of technologies involved.
If you are an expert or just want to study something new while contributing,
you may find interesting opportunities in our project.

== Where to contribute?

The Jenkins project is spread across multiple organizations on GitHub.
You are welcome to contribute to **any** repository in **any** of those organizations, or to any other Jenkins-related repository on GitHub.

* https://github.com/jenkinsci[jenkinsci] - Main organization. Jenkins core, plugins and libraries reside there
* https://github.com/jenkins-infra[jenkins-infra] - Jenkins infrastructure, including the website and other services
* https://github.com/stapler/[stapler] - Stapler Web Framework which is currently maintained by the Jenkins community

> * https://github.com/jenkins-zh[jenkins-zh] - organization of the link:/sigs/chinese-localization/[Chinese Localization SIG]

Reviewer: Shouldn't we also add https://github.com/jenkins-docs

Author: Did not know it exists :D 
CC @MarkEWaite what is the plan about it? Should we get it listed?

Reviewer 2: The repositories in the jenkins-docs organization are used for demonstrations and tutorials in the Jenkins documentation.  The Blue Ocean tutorial uses the https://github.com/jenkins-docs/simple-java-maven-app repository./

Various components in Jenkins have differing review and delivery policies,
so please see the repositories for specific contributing guidelines.

=== Plugins

There are more than link:https://plugins.jenkins.io[1500 plugins] in Jenkins,

> and these plugins implement the most of the Jenkins functionality.

Reviewer 2: ```suggestion
and these plugins implement the majority of Jenkins functionality.
```

> Every plugin is a isolated component which can be developed independently of other parts of the project,

Reviewer 2: ```suggestion
Every plugin is an isolated component which can be developed independently of other parts of the project,
```

with help of the APIs and development tools provided by the project.

> The most of Jenkins contributors work on plugins, and it might be the best way to start contributing to the project.

Reviewer 2: ```suggestion
Most Jenkins contributors work on plugins and it is often the best way to start contributing to the project.
```

Here are some documentation links:

* link:/doc/developer/[Jenkins developer documentation]
* https://wiki.jenkins.io/display/JENKINS/Plugin+tutorial[The plugin tutorial will get you started with Jenkins plugin development]
* https://wiki.jenkins.io/display/JENKINS/Hosting+Plugins[The complete guide to hosting and publishing plugins]
* https://wiki.jenkins.io/display/JENKINS/Adopt+a+Plugin[Adopt a plugin]
* https://wiki.jenkins.io/display/JENKINS/Before+starting+a+new+plugin[Advice before creating a new plugin]
* https://wiki.jenkins.io/display/JENKINS/Pull+Request+to+Repositories[How we handle pull requests to plugin repositories]

=== Jenkins core

> Jenkins core is the heart of any Jenkins installation which provides the kernel functionality and extension APIs being used by Jenkins plugins.

Reviewer 2: ```suggestion
Jenkins core is the heart of any Jenkins installation. It provides the kernel functionality and extension APIs used by Jenkins plugins.
```

=== Infrastructure

As an independent open source project, the Jenkins project maintains most of its own infrastructure including services which help to keep the project running. 
The kind of things that fall into "infrastructure" can span from operating virtual machines, containers, configuring network or developing and maintaining project-specific applications to make the development of Jenkins core and plugins more efficient.

Because we strongly believe in Open Source principles, we also apply them to our infrastructure. 
As such we consider ourself as an open infrastructure project where everybody is invited to learn, share, contribute.

See the link:/projects/infrastructure/[Infrastructure sub-project] for more information and contributing guidelines.

=== Other components

There are hundreds of repositories which do not fall under categories below:
libraries, packaging, developer tools, etc.
You can discover these components by looking into link:/sigs[special interest groups] and link:/projects[sub-projects] pages.

////
TODO(oleg_nenashev): Expand this section?
////

== Newcomers

The Jenkins project always welcomes newcomer contributors.

> We maintain lists of newbie-friendly issues which can be taken.

Reviewer 2: ```suggestion
We maintain lists of newbie-friendly issues and encourage new contributors to work on those issues.
```

* link:https://issues.jenkins-ci.org/issues/?jql=labels%20%3D%20newbie-friendly%20and%20status%20in%20(Open%2C%20%22To%20Do%22%2C%20Reopened)[Newbie-friendly issues in Jenkins JIRA]
* link:https://github.com/search?q=org%3Ajenkinsci+org%3Ajenkins-infra+is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22[Good first issues on GitHub]

Useful links:

* link:https://wiki.jenkins.io/display/JENKINS/Beginners+Guide+to+Contributing#BeginnersGuidetoContributing-Areyouinterestedinwritingcode%3F[Beginners Guide to Contributing]
* Presentation: Contributing to Jenkins (link:https://docs.google.com/presentation/d/1JHgVzWZAx95IsUAZp8OoyCQGGkrCjzUd7eblwd1Y-hA/edit?usp=sharing[slides])
* link:/sigs/advocacy-and-outreach/outreach-programs/[Outreach programs] - Google Summer of Code, Outreachy, Hacktoberfest may be a good opportunity to start contributing to the project

== Experienced contributors

Are you an established contributor to Jenkins and looking for a new challenge?
The link:/security#team[Jenkins Security Team] is looking for members willing to help improve Jenkins security.

> Also, there are many link:/sigs[special interest groups] and link:/projects[sub-projects] which look for contributors 

Reviewer 2: ```suggestion
Also, there are many link:/sigs[special interest groups] and link:/projects[sub-projects] which are seeking contributors 
```
